### PR TITLE
Support local publishing in DynamicPublisher

### PR DESCRIPTION
**Pull request by 32bitkid**
needs tests and documentation but wanted to run this idea/implementation past people. This PR add a local publisher option to the DynamicPublisher block, which should be interoperable with the `LocalSubscriber` block.

If this looks like a good idea, then i'll update the tests and docs to cover—in local testing it works.

### CODE REVIEW

File: dynamic_publisher.py [hunk 1]
```diff
@@ -23,18 +25,38 @@ class DynamicPublisher(PubSubConnectivity, TerminatorBlock):
...
         super().__init__()
         self._cache = keydefaultdict(lambda topic: (self.__create_publisher(topic), None))
         self._cache_lock = Lock()
+        self._is_local = False
+        self._local_id = None
+
+    def configure(self, context):
+        super().configure(context)
+        self._is_local = self.is_local()
+        if self.is_local:
```
Comment (mattdodge) [inline dynamic_publisher.py]: This should be `if self._is_local`

**Comment by mattdodge** [comment]
Looks good. We may normally opt for a `LocalDynamicPublisher` block but I think given this is an advanced block combining them into one block is fine.

Please add a test though to test the local publisher aspect of this block.

File: dynamic_publisher.py [hunk 2]
```diff
@@ -23,18 +25,38 @@ class DynamicPublisher(PubSubConnectivity, TerminatorBlock):
...
+        order=0)
 
     ttl = TimeDeltaProperty(
         title="Time-to-live",
         advanced=True,
         order=0,
         default=dict(seconds=600))
+    is_local = BoolProperty(
+        advanced=True,
+        allow_none=True,
```
Comment (tyoungNIO) [inline dynamic_publisher.py]: Does this need to allow `None`? It's a bool, after all, and older configs without this property defined will use the default. Am I misunderstanding the intent?

Comment (32bitkid) [reply dynamic_publisher.py]: >            Exception: If a value has not been set and a default is not defined
            and allow_none is False.

Ahh, I didn't realize that allow_none didn't really do anything if a default had been set. Removing!

**Comment by 32bitkid** [comment]
added tests and docs.

@mattdodge I thought about adding a new `LocalDynamicPublisher` but that felt confusing or at least very verbose. But, I'm not opposed to it if you think its worth it. But, I would—and did at the time—almost prefer if the regular Publisher/Subscribers had a `is_local` property on them to toggle them into local mode; and then the block renderer in the system designer could/would visualize them uniquely/be able to accurately trace through them. I feel like that makes it easier to adjust/swap behavior later on, _.e.g._ i have a service that has like 20 routes going to a Publisher and then decide, crap I really don't want this to publish/leak to my whole system. I know you can manually go in and edit the block configuration, but I dont think there is a clean way to make that change in the system designer. Though maybe that is more of a missing feature (find-and-replace) of the system designer than a really strong opinion about separation of concerns on blocks.

Anyway, thats the argument that I have in my own head about it.

**Comment by 32bitkid** [comment]
i've rebased this onto master to resolve any conflicts, test pass and I think its ready for review/merge.

**Pull request merged**